libindy/src/commands/crypto.rs

```diff
@@ -209,50 +288,368 @@ impl CryptoCommandExecutor {
+            IndyError::CommonError(CommonError::InvalidStructure(format!(
+                "Failed to serialize JWE {}",
+                err
+            )))
+        })
+    }
+
+    pub fn unpack_msg(&self, jwe_json: Vec<u8>, wallet_handle: i32) -> Result<Vec<u8>> {
+        //serialize JWE to struct
+        let jwe_struct: JWE = serde_json::from_slice(jwe_json.as_slice()).map_err(|err| {
+            IndyError::CommonError(CommonError::InvalidStructure(format!(
+                "Failed to deserialize JWE {}",
+                err
+            )))
+        })?;
+        //decode protected data
+        let protected_decoded_vec = base64::decode(&jwe_struct.protected)?;
+        let protected_decoded_str = String::from_utf8(protected_decoded_vec).map_err(|err| {
+            IndyError::CommonError(CommonError::InvalidStructure(format!(
+                "Failed to utf8 encode data {}",
+                err
+            )))
+        })?;
+        //convert protected_data_str to struct
+        let protected_struct: Protected = serde_json::from_str(&protected_decoded_str).map_err(|err| {
+            IndyError::CommonError(CommonError::InvalidStructure(format!(
+                "Failed to deserialize protected data {}",
+                err
+            )))
+        })?;
+
+        //extract recipient that matches a key in the wallet
+        let (recipient, is_auth_recipient) = self._find_correct_recipient(protected_struct, wallet_handle)?;
+
+        //get cek and sender data
+        let (sender_verkey_option, cek) =
+            match is_auth_recipient {
+                true => self._unpack_cek_authcrypt(recipient, wallet_handle),
+                false => self._unpack_cek_anoncrypt(recipient, wallet_handle),
+            }?; //close cek and sender_data match statement
+
+        //decrypt message
+        let message = self.crypto_service.decrypt_ciphertext(
+            &jwe_struct.ciphertext,
+            &jwe_struct.protected,
+            &jwe_struct.iv,
+            &jwe_struct.tag,
+            &cek,
+        )?;
+
+        //serialize and return decrypted message
+        let res = UnpackMessage {
+            message,
+            sender_verkey: sender_verkey_option,
+        };
+
+        return serde_json::to_vec(&res).map_err(|err| {
+            IndyError::CommonError(CommonError::InvalidStructure(format!(
+                "Failed to serialize message {}",
+                err
+            )))
+        });
+    }
+
+    fn _find_correct_recipient(&self, protected_struct: Protected, wallet_handle: i32) -> Result<(Recipient, bool)>{
+        for recipient in protected_struct.recipients {
+            let my_key_res : Result<Key> = self.wallet_service.get_indy_object(
+                wallet_handle,
+                &recipient.header.kid,
+                &RecordOptions::id_value()
+            ).map_err(|err| IndyError::WalletError(err));
+
+
+            if my_key_res.is_ok() {
+                match recipient.header.sender {
```

Comment:
I suggest to replace this `match` by something like
`Ok((recipient, recipient.header.sender.is_some()))`

Reply:
added this suggestion